### PR TITLE
refactor(side-menu): reworked the side-menu

### PR DESCRIPTION
**Pull request by victorscania**
**Describe pull-request**  
remake of the side-menu

**Solving issue**  
Fixes: [AB#1625](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1625)

**How to test**  
run storybook and go to side-menu

### CODE REVIEW

**Comment by theJohnnyMe** [comment]
I really like how much smooth it feels now in the Storybook example. I will check the code later today or tomorrow morning first thing. 
 @alexlioliopoulos please take a look too.

**Review by alexlioliopoulos — CHANGES_REQUESTED** [review]
As smooth as Snoop Dog indeed!

A couple of things I am noticing:

- It seems that there is no focused state (blue outline when you use tab to navigate)

- Submenu item text does not seem to be in the centre.
image

- A small one which I am not even sure if it's solvable. When we have a scrollbar, it kinda creates some space for the scrollbar to exist, which however makes the dividers feel like floating, since they are not touching the verical line.

image

**Comment by areblk** [comment]
## Comments on stories
1. Can we have a story for a non-collapsible side menu that has no prefix icons? I'd like to verify the margins when there are no icons.
2. Non-collapsible should be first story, since it's default side menu
3. _Probably out of scope. (ping @alexlioliopoulos )_ Fullpage pattern stories now completely broken due to breaking changes for both header and side menu. It's located in \sdds\components\src\components\theme\fullpage.stories.js.

## Other
4. All sub items 3 animate on load. Intentional? Looks a bit weird since they're the only item labels that do this. See video

https://user-images.githubusercontent.com/97448894/154299445-828134ce-36d6-4848-8c33-f5d93e4ef940.mp4


## Deviations
5. Alex also mentioned this. Sub item labels not equal top and bottom margin (center aligned), see screenshot
6. Sub item total height should be 48 px, not 56, see screenshot
Sub item margin

Design here: https://dw1u92ia6bs84.cloudfront.net/Images/3b503de9f50b460493ca12ee91c3f509.png
https://dw1u92ia6bs84.cloudfront.net/Images/bd49da3168744e88b123d6fcf1dee86b.png

7. Side menu shouldn't collapse to menu drawer for lg viewport, but it shouldn't. 
doesn't show on lg viewport

8. Collapsed side menu, Sub menu-label should show on hover, but only shows sub page links
9. Alex also mentioned this. Missing border and dropshadow of pop-over, see screenshot 
pop-over misses borders and shadow

Design: https://dw1u92ia6bs84.cloudfront.net/Images/Image_feac7ae447.png

**Comment by alexlioliopoulos** [comment]
@victorscania 
Another thing that I'd like to comment on. It might already be taken care of, but I can't be sure from the StoryBook example. Ignore this message if it's not applicable!
So, when a sub-item has been selected, but you then collapse the dropdown it lives under, then this collapsed dropdown needs to be in "selected" state.

image
image

**Review by gaonita — COMMENTED** [review]
Great work! huge task, it took long time for me to review. 
and super nice to remove all the complex '_navbar' scss files.
- I left some comments

- Bug: I found that on mobile screen, the collapse button is removed but the space still takes up the height and eventually it blocks the menu items like this screenshot. 

Screenshot 2022-02-18 at 18 49 31

- probably you already know; you can find the current side menu in our angular and react demo, not fully updated but in case you need reference of the basic behaviour.

File: components/src/patterns/side-menu/side-menu.stories.js [hunk 1]
```diff
@@ -34,245 +20,302 @@ const Template = (args) => {
...
+Collapsible.args = {
+  collapsible: true,
 };
 
-export const Dropdown = Template.bind({});
+export const Non_collapsible = Template.bind({});
 
-Dropdown.args = {
-  dropdown: true,
+Non_collapsible.args = {
```
[the remark this answers is not in the document]
Comment (areblk) [reply components/src/patterns/side-menu/side-menu.stories.js:319]: I vote that Side Menu stories should be in this order: 1. Non-collapsible without icon, 2. Non-collapsible with icon, 3. Collapsible with icon

**Comment by areblk** [comment]
## Not resolved
- (8) Collapsed menu, hovering icon of sub-menu should show sub menu-item on top of sub page items. https://dw1u92ia6bs84.cloudfront.net/Images/Image_feac7ae447.png

## New

10. Name parts in side menu same way as in design so it's easier to talk about/refer to. Remove numbering of pages (not needed past unit testing). Design reference: https://dw1u92ia6bs84.cloudfront.net/Images/ae84d63ed8084be699a3d47f7b04ae30.png
- item --> **Page link** or **Sub-menu**. (item 1, item 4, item 6-8 --> Page link; A page name that has a very long name --> Page link with a very long name; item 2-3 --> Sub-menu)
- Sub item X --> **Sub Page name**
11. Wrong colour on container when item is hovered or selected. Should be $grey-50 according to Color table, https://digitaldesign.scania.com/components/sidemenu/style#colour Please note that pressed items container has different colour values if they're popover items ($grey-100) or not ($grey-300)
12. Container should change colour on focus to $grey-50

## To be defined
13. Missing bottom border when sub page-items are focused. _Although_, if it should have a border at all is unclear because of *inconclusive styling guidelines*. Table say it should look exactly like the selected state, but image example state otherwise. I've asked the designers what it should look like (@alexlioliopoulos). https://digitaldesign.scania.com/components/sidemenu/style

## Resolved since last commit:
- stories and order (1, 2)
- onload animation removed (4)
- sub item labels placement (5)
- sub item height (6) - false negative due to misreading of storybook add-on, sorry about that!
- responsive behaviour on lg viewport (7)
- dropshadow and border on pop-over (9)

**Comment by areblk** [comment]
### New
14. (somewhat similar to 13) Missing top-border when first page link in side menu is focused (only visible on mobile menu)
image
15. When side menu is collapsed, Sub Pages shouldn't be included in tab-index. 
16. When side menu is collapsed, keyboard focusing on a Sub-menu item and pressing enter should open the pop-over menu and make that able to tab through. Although, this might clash the design spec. saying that pop-over menu should only show on hover. 
17. **EDGE CASE** When collapsed side menu is scrollable, it's somewhat hard to focus (with a mouse) on the pop-over Sub-menu and it's Sub Pages. I think this might be a very uncommon state and therefore not worth putting too much effort on in fixing. Or what do you say @alexlioliopoulos 

Also, focusing and selecting elements with keyboard in expanded side menu works like a charm! (ping @theJohnnyMe)

### Resolved since last review
- (8) Sub-menu item appear on hover
- (10) Naming of parts
- (11, 12) Color on hover 
- (13) Bottom border on focused sub menu-item

**Comment by alexlioliopoulos** [comment]
> ### New
> 14. (somewhat similar to 13) Missing top-border when first page link in side menu is focused (only visible on mobile menu)
>     image
> 15. When side menu is collapsed, Sub Pages shouldn't be included in tab-index.
> 16. When side menu is collapsed, keyboard focusing on a Sub-menu item and pressing enter should open the pop-over menu and make that able to tab through. Although, this might clash the design spec. saying that pop-over menu should only show on hover.
> 17. **EDGE CASE** When collapsed side menu is scrollable, it's somewhat hard to focus (with a mouse) on the pop-over Sub-menu and it's Sub Pages. I think this might be a very uncommon state and therefore not worth putting too much effort on in fixing. Or what do you say @alexlioliopoulos
> 
> Also, focusing and selecting elements with keyboard in expanded side menu works like a charm! (ping @theJohnnyMe)
> 
> ### Resolved since last review
> * (8) Sub-menu item appear on hover
> * (10) Naming of parts
> * (11, 12) Color on hover
> * (13) Bottom border on focused sub menu-item

No17 is not such an edge case for me, and would like to have it solved if it's not too much trouble.

**Comment by alexlioliopoulos** [comment]
One thing that I'd like to ask from my side is the following:

If a sub-item is selected, according to design, the parent item should be in selected state. In our current example this is not the case, but that might be because we kinda have 2 things selected at the same time.

But let's say the 4th box was not selected, but only the sub-item under the 3rd box was selected. In this case, would the 3rd box show as selected or not?

image

File: components/src/patterns/side-menu/side-menu.scss [hunk 2]
```diff
@@ -2,291 +2,398 @@
...
+      &::after {
+        content: ' ';
+        position: absolute;
+        background: var(--sdds-black);
+        opacity: 0;
+        pointer-events: none;
+        top: 0;
+        bottom: 0;
+        left: 100%;
+        width: 1000px;
```
[the remark this answers is not in the document]
Comment (areblk) [reply components/src/patterns/side-menu/side-menu.scss:103]: ,I think no, since 100vw produces overflow on Windows :-(

**Review by areblk — APPROVED** [review]

**Comment by theJohnnyMe** [comment]
The branch is rebased with the latest master.

File: components/src/patterns/side-menu/side-menu.scss [hunk 3]
```diff
@@ -2,291 +2,398 @@
...
+        padding: 24px 16px 24px 24px;
+        display: flex;
+        align-items: center;
+        position: relative;
+        min-height: 20px;
+        box-sizing: content-box;
+        white-space: nowrap;
+        transition: background-color 0.15s ease-in-out;
+
+        span {
```
[the remark this answers is not in the document]
Comment (areblk) [reply components/src/patterns/side-menu/side-menu.scss:222]: Remember to also update documentation and code examples with this change

**Comment by areblk** [comment]
- Note to self. Johnnys wishlist to Azure for future improvement: animate dropdown list opening (sub-menu) so that it opens smoother, like the movement that the collapse item does when it's being collapsed and non-collapsed

File: components/src/patterns/side-menu/side-menu.scss [hunk 4]
```diff
@@ -2,291 +2,398 @@
...
       }
+    }
+
+    &__icon {
+      margin-right: 12px;
+      line-height: 0;
+      width: 20px;
+      height: 20px;
+      min-width: 20px;
+      transition: 0.15s ease-in-out;
```
[the remark this answers is not in the document]
Comment (areblk) [reply components/src/patterns/side-menu/side-menu.scss:327]: not now, future improvement

File: components/src/patterns/side-menu/side-menu.stories.js [hunk 5]
```diff
@@ -34,245 +20,302 @@ const Template = (args) => {
     .sb-show-main.sb-main-padded {
       padding: 0;
     }
-    .sdds-demo-container { 
+    .sdds-demo-container {
```
[the remark this answers is not in the document]
Comment (areblk) [reply components/src/patterns/side-menu/side-menu.stories.js:25]: johnny and I checked and it's needed for this story

**Review by areblk — APPROVED** [review]

**Pull request merged**